lib/reader.js

```diff
@@ -24,66 +24,64 @@ class Reader {
 	}
 
 	readString() {
-		var b = this.buffer[this.position++];
-		
-		if ((b & 0x80) == 0) {
-			this.position--;
-			return this.readAscii(); // ASCII.
+		var peek = this.buffer[this.position];
+
+		if ((peek & 0x80) == 0) {
+			return this.readAsciiString();
+		} else {
+			return this.readUtf8String();
 		}
-		
-		// Null, empty, or UTF8.
-		var charCount = this.readUtf8Length(b);
-		switch (charCount) {
-		case 0:
+	}
+
+	readUtf8String() {
+		var charCount = this.readUtf8Length();
+		// length == 0 is a special code for "null"
+		if (charCount == 0) {
 			return null;
-		case 1:
-			return "";
+		} else {
+			return this.readUtf8(charCount - 1);
 		}
-		charCount--;
-		return this.readUtf8(charCount);
 	}
 
-	readUtf8Length(b) {
-		var result = b & 0x3F; // Mask all but first 6 bits.
-
+	readUtf8Length() {
 		// bit 8 is the utf marker
 		// bit 7 is the length marker *for the first byte*
 		// then it's bit 8 again
 
-		// just short circuit if there's no continuation
-		if ((b & 0x40) == 0) {
-			return result;
-		}
-
-		b = this.buffer[this.position++];
-		result |= (b & 0x7F) << 6;
+		var b = this.buffer[this.position++];
+		var result = b & 0x3F; // Mask all but first 6 bits.
+		var cont = b & 0x40;
 
-		var i = 1;
-		while ((b & 0x80) != 0) {
+		var i = 0;
+		while (cont != 0 && this.position < this.buffer.length) {
 			b = this.buffer[this.position++];
 			result |= (b & 0x7F) << (6 + (7 * i++));
+			cont = b & 0x80;
 		}
-				
+
 		return result;
 	}
 
 	readUtf8(charCount) {
 		let decoder = new StringDecoder("utf8");
 		var result = "";
 
-		while (result.length < charCount) {
-			result += decoder.write(this.buffer.slice(this.position++, this.position));
+		while (result.length < charCount && this.position < this.buffer.length) {
+			result += decoder.write(this.buffer.slice(this.position, this.position+1));
+			this.position++;
```

Comment:
I'm not a fan of using `++` in the middle of an expression at the best of times, but I'm super not a fan of using it and then also using the same variable again in the same expression...